### ISSUE TITLE
Unexpected exception <ssl.SSLError> instead of <ConnectionClosedError> when 'max_size' is exceeded

### ISSUE THREAD

**Issue by etkaar** [aaugustin/websockets#808]
When I send larger packets (> 1 MiB) to the WebSocket server, I just get this error, nothing more and the connection to the client closes:

```
SSL error in data received
protocol: <asyncio.sslproto.SSLProtocol object at 0x7f161215be80>
transport: <_SelectorSocketTransport fd=9 read=polling write=<idle, bufsize=0>>
Traceback (most recent call last):
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 526, in data_received
    ssldata, appdata = self._sslpipe.feed_ssldata(data)
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 207, in feed_ssldata
    self._sslobj.unwrap()
  File "/usr/lib/python3.7/ssl.py", line 767, in unwrap
    return self._sslobj.shutdown()
ssl.SSLError: [SSL: KRB5_S_INIT] application data after close notify (_ssl.c:2609)
```

This is due to a limit which can be fixed by changing `max_size` in **serve()**:

`websockets.serve(self.websocketClientHandler, host, port, max_size=(5 * 1024 * 1024))`

However, this does not explain this error message, there should be thrown an Exception.

According to this, a `ConnectionClosedError` exception is expected:
https://websockets.readthedocs.io/en/stable/api.html?highlight=websockets.serve#websockets.protocol.WebSocketCommonProtocol

**Comment by aaugustin** [comment]
If you [enable debug logs](https://websockets.readthedocs.io/en/stable/cheatsheet.html#debugging) you should see a connection failure with code 1009, which is "payload too big" (I don't expect you to know this).

At that point, websockets closes the connection. If the other side keeps sending data, this would explain the error message.

Could you confirm what client you're using to connect to the websockets server?

**Comment by etkaar** [comment]
> Could you confirm what client you're using to connect to the websockets server?
I use Chromium for the client :)

**Comment by aaugustin** [comment]
That rules out a client bug. This shouldn't be too hard to reproduce, then.

**Comment by etkaar** [comment]
Well, I intentionally sent a payload > 1 MiB to test what would happen then on server side. I added now following to the beginning of my server program:

```
import logging
logger = logging.getLogger('websockets')
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler())
```

However, still only this error message from the first post appears, no full traceback, so I am not able to find out from where it comes.

**Comment by aaugustin** [comment]
No doubt this error should be reported better. Hence, the "bug" label.

**Comment by etkaar** [comment]
Ah, I see. But what is the reason there is no exception thrown which I could just catch?

**Comment by aaugustin** [comment]
If you call `recv()` in your connection handler while or after this happens, then `recv()` should raise an exception.

**Comment by etkaar** [comment]
This is, by the way, not the only error message.

Sometimes I receive – for no obvious reason – messages like that. But all works as intended, so I cannot assign that at all to any misfunction. Also that I get **VERSION_TOO_LOW** is very strange:

```
SSL handshake failed
protocol: <asyncio.sslproto.SSLProtocol object at 0x7f8d9035d0f0>
transport: <_SelectorSocketTransport fd=10 read=polling write=<idle, bufsize=0>>
Traceback (most recent call last):
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 625, in _on_handshake_complete
    raise handshake_exc
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: VERSION_TOO_LOW] version too low (_ssl.c:1056)
SSL error in data received
protocol: <asyncio.sslproto.SSLProtocol object at 0x7f8d9035d0f0>
transport: <_SelectorSocketTransport closing fd=10 read=idle write=<idle, bufsize=0>>
Traceback (most recent call last):
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 526, in data_received
    ssldata, appdata = self._sslpipe.feed_ssldata(data)
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: VERSION_TOO_LOW] version too low (_ssl.c:1056)
SSL handshake failed
protocol: <asyncio.sslproto.SSLProtocol object at 0x7f8d9035d400>
transport: <_SelectorSocketTransport fd=10 read=polling write=<idle, bufsize=0>>
Traceback (most recent call last):
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 625, in _on_handshake_complete
    raise handshake_exc
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: NO_SHARED_CIPHER] no shared cipher (_ssl.c:1056)
SSL error in data received
protocol: <asyncio.sslproto.SSLProtocol object at 0x7f8d9035d400>
transport: <_SelectorSocketTransport closing fd=10 read=idle write=<idle, bufsize=0>>
Traceback (most recent call last):
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 526, in data_received
    ssldata, appdata = self._sslpipe.feed_ssldata(data)
  File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: NO_SHARED_CIPHER] no shared cipher (_ssl.c:1056)
```

**Comment by aaugustin** [comment]
Ah, httpx and aiohttp had the same issue and a maintainer of httpx investigated it thoroughly:
https://github.com/aio-libs/aiohttp/issues/3535#issuecomment-544279431

Calling wait_closed on the stream writer could do the trick:
https://docs.python.org/3/library/asyncio-stream.html?highlight=wait_closed#asyncio.StreamWriter.wait_closed

**Comment by aaugustin** [comment]
Except websockets doesn't use a stream writer...

This was supposed to be fixed in Python 3.7.4, which was released one year before you filed this ticket.

Could you confirm which Python version you're using? If you're using 3.7 - 3.7.3, could you upgrade to 3.7.4 and see if that helps?

**Comment by aaugustin** [comment]
Please let me know if you still see this on Python ≥ 3.7.4 and I'll reopen the issue.
